### DISCUSSION TITLE
[Spec]  Task Api First

### DISCUSSION THREAD

**Discussion by pictos** [dotnet/maui#133] [General]
I don't know if this name represents my proposal, feel free to change it. (I'm terrible naming things)
# Change some events and methods to `Task/Task<T>`

Today is practically impossible to have a real app without async code (aka. `Task` and `Task<T>`) and the XF has a lot of APIs that weren't built in this way and, sometimes, we have a hard time to use them with async code.

My proposal is to change the return of those APIs to `Task` or `Task<T>` (when it's possible). To illustrate better this scenario I can use the `OnBackButtonPressed`, it's a virtual method inside the `Page` class.

```csharp
// Inside the class Page
protected virtual bool OnBackButtonPressed ();
```
Since this method returns a `bool` it's very hard to do some async operation inside it. Also with `void` methods, this can be tricky, for example in Shell the `OnNavigating` event, if you want to show a pop-up to ask the user if he wants to leave the page.

```csharp
async void OnShellNavigating(object sender, ShellNavigatingEventArgs e)
{
    var current = e.Current;
    if (current.Location.OriginalString.Contains("MainViewModel"))
    {
        var result = await DisplayAlert("Alert", "Are you sure?", "Yes", "No");
        if (!result)
                e.Cancel();
    }
}
```
In this example, when the DisplayAlert appears, the navigation occurs, so we don't have the user's input.

# API
Here I'll list some methods/events that would be good to have async support. For events, maybe we can use `Func<Task>`.

| API | Description |
| ------------- | ------------- |
| OnAppearing | Page |
| OnDisappearing| Page |
| OnBackButtonPressed | Page |
| OnResume | Application |
| OnStart | Application |
| OnSleep | Application |
| OnNavigating | Shell |
| OnNavigated | Shell |
 
# Scenarios

## C# Example
```csharp
async Task OnShellNavigating(object sender, ShellNavigatingEventArgs e)
{
    var current = e.Current;
    if (current.Location.OriginalString.Contains("MainViewModel"))
    {
        var result = await DisplayAlert("Alert", "Are you sure?", "Yes", "No");
        if (!result)
                e.Cancel();
    }
}
```

# Difficulty : [low/medium/high]
Medium

**Comment by dishantmshah** [comment]
I agree with this point, but when you look at the API's which you have mentioned most of them are top event handlers and it's a good practice to use `void` for those Event Handlers. In case if you want to do any async operation you can still do it inside those methods and there are a number of alternatives to do that. One approach would be to use `TaskCompletionSource` or `ContinueWith` .

**Comment by pictos** [comment]
@dishantmshah, the handler of those events can be `Tasks`, for example, the `OnResume` inside the platform-specific can be an `async void` method, but the implementation inside the Application class can be a `Task`. I know that it's not possible for every API but it will be good if we try to build APIs thinking in async behavior.

**Comment by dhindrik** [comment]
It can properly affect performance in a bad way and it is pretty easy to make it async on your own.

**Comment by pictos** [comment]
@dhindrik,

> It can properly affect performance in a bad way

I agree with you on this point, but I think it's for a good reason, as I said actual apps always will need to do async operations. As you can see [here](https://devblogs.microsoft.com/premier-developer/the-performance-characteristics-of-async-methods/) Tasks can be 15% slower than sync methods. But, maybe, we can use the `ValueTasks` in some points to avoid this hit in every API.

> it is pretty easy to make it async on your own.

I disagree with you here, to me, pretty easy means add the async/await keyword and everything works like a charm, sadly this isn't always true, for some scenarios I'll use those ones that I used on my spec.

The first one is `protected virtual bool OnBackButtonPressed ();`, this method returns a bool, so I can't easily change this to async. I'll need to do more work to archive that.

The second one is Shell.Navigating event, here it's simple (at first glance) I just need to add the async and done, right? No. Let's take a look at this.

```csharp
async void OnShellNavigating(object sender, ShellNavigatingEventArgs e)
{
        var result = await DisplayAlert("Alert", "Are you sure?", "Yes", "No");
        if (!result)
            e.Cancel();
}
```
Reading this code, you can assume that if the user taps on the "No" button option, the navigation doesn't occur. But the practical behavior is different when the DisplayAlert appears, the rest of the code runs, the navigation occurs and the user response doesn't have any effect.

As you can see some scenarios aren't _pretty easy_ to make it async. Maybe I don't see an easy way here, because I'm missing something, if so please let me know.
